### PR TITLE
Various bug fixes

### LINKED ISSUE #940: While updating product Class 'Illuminate\Support\Facades\Input' not found

### LINKED ISSUE #1020: When change update expire date from order page loader is overlapping.
When change update expiry date from the order page loader is overlapping.
Please see the screen shot:-
Annotation 2020-04-22 121106

### LINKED ISSUE #1036: Fix installer issues

### LINKED ISSUE #1037: Link Order on Invoice Datatable and view invoice page

### LINKED ISSUE #1042: Label for partial payment if complete payment is not done

### LINKED ISSUE #1043: If a product is renewed, show a label that shows that it is renewed

### LINKED ISSUE #1044: 2fa icon on clients datatable

### LINKED ISSUE #1046: Link Invoice in payment tab with invoice

### LINKED ISSUE #1053: Some time when user login it's redirecting to /home page.
Annotation 2020-05-02 145909

### LINKED ISSUE #1055: Send Email after successful and failed payment

### PR DESCRIPTION
**Pull request by ashu555**
Fixes #1020 
Fixes #940 
Fixes #1037 
Fixes #1042 
Fixes #1043 
Fixes #1046 
Fixes #1036 
Fixes #1044
Fixes #1053 
Fixes #1055 

Run php artisan migrate -- path="database\migrations\2020_04_30_173907_add_is_renewed_to_invoices.php"

### CODE REVIEW

File: app/Http/Controllers/DashboardController.php [hunk 1]
```diff
@@ -302,15 +302,14 @@ public function getRecentInvoices()
         return Invoice::with("user:id,first_name,last_name,email,user_name")
             ->leftJoin("currencies", "invoices.currency", "=", "currencies.code")
             ->leftJoin("payments", "invoices.id", "=", "payments.invoice_id")
-            ->select("invoices.id as invoice_id", "invoices.number as invoice_number", "invoices.grand_total",
+            ->select("invoices.id as invoice_id", "invoices.number as invoice_number", "invoices.grand_total", "invoices.status as status",
```
Comment (avinash403) [inline app/Http/Controllers/DashboardController.php:305]: no need to write `as status`, since column name is status, it will come as status

File: app/Http/Controllers/User/AdminOrderInvoiceController.php [hunk 2]
```diff
@@ -78,10 +86,39 @@ public function getClientInvoice($id)
...
+        if($allInvoicesRelatedToOrder) {
+        $orderid = ($allInvoicesRelatedToOrder[0]);
+        $order = Order::find($allInvoicesRelatedToOrder[0]);
+        return '<a href='.url('orders/'.$orderid).'>'. $order->number.'</a>';
+       } else {
+        return '--';
+       }
+    }
+
+    public static function getStatusLabel($status, $badge='label')
```
Comment (avinash403) [inline app/Http/Controllers/User/AdminOrderInvoiceController.php:105]: looks useful. Please move to helpers

File: app/Http/Controllers/Front/ClientController.php [hunk 3]
```diff
@@ -384,6 +389,7 @@ public function getOrder($id)
                 compact('invoice', 'order', 'user', 'plan', 'product', 'subscription', 'licenseStatus', 'installationDetails', 'allowDomainStatus')
             );
         } catch (Exception $ex) {
+            dd($ex);
```
Comment (avinash403) [inline app/Http/Controllers/Front/ClientController.php:392]: remove dd

File: app/Http/Controllers/Order/InvoiceController.php [hunk 4]
```diff
@@ -209,14 +207,14 @@ public function getInvoices(Request $request)
     public function show(Request $request)
     {
         try {
-            $id = $request->input('invoiceid');
-            $invoice = $this->invoice->where('id', $id)->first();
-            $invoiceItems = $this->invoiceItem->where('invoice_id', $id)->get();
+            $invoice = $this->invoice->where('id', $request->input('invoiceid'))->first();
```
Comment (avinash403) [inline app/Http/Controllers/Order/InvoiceController.php:210]: Please use select query if you need only few columns

File: app/Http/Controllers/Order/InvoiceController.php [hunk 5]
```diff
@@ -209,14 +207,14 @@ public function getInvoices(Request $request)
...
+            $invoiceItems = $this->invoiceItem->where('invoice_id', $request->input('invoiceid'))->get();
```
Comment (avinash403) [inline app/Http/Controllers/Order/InvoiceController.php:211]: Please use select query if you need only few columns

File: app/Http/Controllers/Order/OrderController.php [hunk 6]
```diff
@@ -20,6 +20,7 @@
 class OrderController extends BaseOrderController
 {
     // NOTE FROM AVINASH: utha le re deva
+    // NOTE: don't lose hope.
```
Comment (avinash403) [inline app/Http/Controllers/Order/OrderController.php:23]: 👍

File: app/Http/Controllers/Product/BaseProductController.php [hunk 7]
```diff
@@ -288,18 +288,14 @@ public function getPrice(Request $request)
...
-            if (\Input::has('github_owner') && \Input::has('github_repository')) {
-                $owner = \Input::get('github_owner');
-                $repo = \Input::get('github_repository');
-                $product = Product::find($productid);
-                $github_controller = new \App\Http\Controllers\Github\GithubController();
-                $version = $github_controller->findVersion($owner, $repo);
-                $product->version = $version;
-                $product->save();
-            }
+            $product = Product::find($productid);
```
Comment (avinash403) [inline app/Http/Controllers/Product/BaseProductController.php:294]: please use select statement

File: app/Http/Controllers/Product/ExtendedBaseProductController.php [hunk 8]
```diff
@@ -175,10 +175,9 @@ public function adminDownload($id, $invoice = '', $api = false)
      */
     public function saveCartValues($input, bool $can_modify_agent, bool $can_modify_quantity)
     {
-        $this->product->show_agent = $input['show_agent'] == 1 ? 1 : 0; //if Show Agents Selected
+        $this->product->show_agent = $input['show_agent'] == 1 ?: 0; //if Show Agents Selected
```
Comment (avinash403) [inline app/Http/Controllers/Product/ExtendedBaseProductController.php:178]: only
```
 $this->product->show_agent = $input['show_agent'] == 1
```
will do the same thing as the entire statement, plus less processing power needed

File: app/Http/Controllers/User/AdminOrderInvoiceController.php [hunk 9]
```diff
@@ -19,15 +20,21 @@ public function getClientInvoice($id)
...
-                            $tz = \Auth::user()->timezone()->first()->name;
-                            $date1->setTimezone(new \DateTimeZone($tz));
-                            $date = $date1->format('M j, Y, g:i a ');
-
-                            return $date;
+                            return getDateHtml($model->date);
                         })
                         ->addColumn('invoice_no', function ($model) {
-                            return      '<a href='.url('invoices/show?invoiceid='.$model->id).'>'.$model->number.'</a>';
+                            if($model->is_renewed) {
```
Comment (avinash403) [inline app/Http/Controllers/User/AdminOrderInvoiceController.php:26]: can be written like 
```
$label = "<a href='.url('invoices/show?invoiceid='.$model->id).'>'.$model->number.'</a>";
if($model->is_renewed){
    return $label. '<br>'.$this->getStatusLabel('renewed');
}
return $label;
```

File: app/Http/Controllers/User/AdminOrderInvoiceController.php [hunk 10]
```diff
@@ -78,10 +86,39 @@ public function getClientInvoice($id)
...
                                     ."   $editAction $action";
                         })
-                         ->rawColumns(['checkbox', 'date', 'invoice_no', 'total', 'paid', 'balance', 'status', 'action'])
+                         ->rawColumns(['checkbox', 'date', 'invoice_no', 'order_no', 'total', 'paid', 'balance', 'status', 'action'])
                         ->make(true);
     }
 
+    public static function getOrderLink($invoice)
+    {
+        $allInvoicesRelatedToOrder = $invoice->orderRelation()->pluck('order_id')->toArray();
```
Comment (avinash403) [inline app/Http/Controllers/User/AdminOrderInvoiceController.php:95]: please use `->value('order_id')`. If value doesn't exist, it will give `null`. Using $allInvoicesRelatedToOrder[0] isn't required

File: app/Http/Controllers/User/AdminOrderInvoiceController.php [hunk 11]
```diff
@@ -78,10 +86,39 @@ public function getClientInvoice($id)
                             style='color:white;'> </i>&nbsp;&nbsp;View</a>"
                                     ."   $editAction $action";
                         })
-                         ->rawColumns(['checkbox', 'date', 'invoice_no', 'total', 'paid', 'balance', 'status', 'action'])
+                         ->rawColumns(['checkbox', 'date', 'invoice_no', 'order_no', 'total', 'paid', 'balance', 'status', 'action'])
                         ->make(true);
     }
 
+    public static function getOrderLink($invoice)
```
Comment (avinash403) [inline app/Http/Controllers/User/AdminOrderInvoiceController.php:93]: I believe this is getting used in a loop. It will kill performance. Please use joins in the main query

File: app/Http/Controllers/User/AdminOrderInvoiceController.php [hunk 12]
```diff
@@ -138,14 +170,10 @@ public function getPaymentDetail($id)
                             value=".$model->id.' name=select[] id=checkpayment>';
                         })
                         ->addColumn('invoice_no', function ($model) {
-                            return $model->invoice()->first() ? $model->invoice()->first()->number : '--';
+                            return $model->invoice()->first() ? '<a href='.url('invoices/show?invoiceid='.$model->invoice()->first()->id).'>'.$model->invoice()->first()->number.'</a><br>' : '--';
```
Comment (avinash403) [inline app/Http/Controllers/User/AdminOrderInvoiceController.php:173]: use `->count()` instead. It queries all the data from the database just for checking its existence.

File: app/Http/Controllers/User/ClientController.php [hunk 13]
```diff
@@ -84,21 +84,29 @@ public function getClients(Request $request)
...
                             } else {
-                                $email = "<span class='glyphicon glyphicon-envelope'
-                                 style='color:red' title='unverified email'></span>";
+                                $email =  "<span class='glyphicon glyphicon-envelope'  style='color:red'  <label data-toggle='tooltip' style='font-weight:500;' data-placement='top' title='Unverified email'>
+                                    </label></span>";
                             }
                             if ($model->mobile_verified == 1) {
-                                $mobile = "<span class='glyphicon glyphicon-phone' 
-                                style='color:green' title='verified mobile'></span>";
+                                $mobile = "<span class='glyphicon glyphicon-phone'  style='color:green'  <label data-toggle='tooltip' style='font-weight:500;' data-placement='top' title='Unverified mobile'>
```
Comment (avinash403) [inline app/Http/Controllers/User/ClientController.php:95]: please find some way of reusing it

File: resources/views/themes/default1/front/clients/show-invoice.blade.php [hunk 14]
```diff
@@ -75,7 +75,9 @@
                             </div><!-- /.col -->
                             <div class="col-sm-4 invoice-col">
                                 <b>Invoice   #{{$invoice->number}}</b><br/>
-                               
+                                @if($invoice->order()->first())
+                                <b>Order</b>   &nbsp; # <a href="{{ url('my-order/'.$invoice->order()->first()->id) }}">{{$invoice->order()->first()->number}}</a>
```
Comment (avinash403) [inline resources/views/themes/default1/front/clients/show-invoice.blade.php:79]: please use ->value

File: resources/views/vendor/installer/environment-wizard.blade.php [hunk 15]
```diff
@@ -512,7 +504,31 @@
...
+<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js" type="text/javascript"></script>
     <script type="text/javascript">
+        $("#setup").on('submit',function(e) {
+        $("#submitme").attr("disabled", true);
+        $('#submitme').val('Installing, please wait...');
+        })
+
+
+        $('#app_name').keyup(function(e){
+            if(e.keyCode == 32) {
```
Comment (avinash403) [inline resources/views/vendor/installer/environment-wizard.blade.php:516]: what is this for?

File: vendor/rachidlaasri/laravel-installer/src/Helpers/DatabaseManager.php [hunk 16]
```diff
@@ -44,8 +44,7 @@ private function migrate($outputLog)
             Artisan::call('migrate', ["--force"=> true], $outputLog);
         }
         catch(Exception $e){
-             throw new Exception('Cannot Establish Database Conection.Check Details');
-            // return $this->response($e->getMessage());
+             throw new Exception($e->getMessage());
```
Comment (avinash403) [inline vendor/rachidlaasri/laravel-installer/src/Helpers/DatabaseManager.php:47]: its inside vendor. Are you sure about it?

File: app/Http/Controllers/DashboardController.php (hunk 1 above)
Comment (ashu555) [reply app/Http/Controllers/DashboardController.php:305]: Fixed

File: app/Http/Controllers/Front/ClientController.php (hunk 3 above)
Comment (ashu555) [reply app/Http/Controllers/Front/ClientController.php:392]: Fixed

File: app/Http/Controllers/Order/InvoiceController.php (hunk 4 above)
Comment (ashu555) [reply app/Http/Controllers/Order/InvoiceController.php:210]: fixed

File: app/Http/Controllers/Order/InvoiceController.php (hunk 5 above)
Comment (ashu555) [reply app/Http/Controllers/Order/InvoiceController.php:211]: Fixed

File: app/Http/Controllers/Product/BaseProductController.php (hunk 7 above)
Comment (ashu555) [reply app/Http/Controllers/Product/BaseProductController.php:294]: Fixed

File: app/Http/Controllers/Product/ExtendedBaseProductController.php (hunk 8 above)
Comment (ashu555) [reply app/Http/Controllers/Product/ExtendedBaseProductController.php:178]: fixed

File: app/Http/Controllers/User/AdminOrderInvoiceController.php (hunk 9 above)
Comment (ashu555) [reply app/Http/Controllers/User/AdminOrderInvoiceController.php:26]: Fixed

File: app/Http/Controllers/User/AdminOrderInvoiceController.php (hunk 10 above)
Comment (ashu555) [reply app/Http/Controllers/User/AdminOrderInvoiceController.php:95]: fixed

File: app/Http/Controllers/User/AdminOrderInvoiceController.php (hunk 11 above)
Comment (ashu555) [reply app/Http/Controllers/User/AdminOrderInvoiceController.php:93]: fixed

File: app/Http/Controllers/User/AdminOrderInvoiceController.php (hunk 12 above)
Comment (ashu555) [reply app/Http/Controllers/User/AdminOrderInvoiceController.php:173]: Fixed

File: app/Http/Controllers/User/ClientController.php (hunk 13 above)
Comment (ashu555) [reply app/Http/Controllers/User/ClientController.php:95]: fixed

File: resources/views/themes/default1/front/clients/show-invoice.blade.php (hunk 14 above)
Comment (ashu555) [reply resources/views/themes/default1/front/clients/show-invoice.blade.php:79]: Fixed

File: resources/views/vendor/installer/environment-wizard.blade.php (hunk 15 above)
Comment (ashu555) [reply resources/views/vendor/installer/environment-wizard.blade.php:516]: Space bar ASCII code is 32. We are disallowing space in App name in installer as .env don't accept those

File: vendor/rachidlaasri/laravel-installer/src/Helpers/DatabaseManager.php (hunk 16 above)
Comment (ashu555) [reply vendor/rachidlaasri/laravel-installer/src/Helpers/DatabaseManager.php:47]: Yes, later we can improve the installer and handle all functions from Billing's core files

File: app/Http/Controllers/Order/ExtendedBaseInvoiceController.php [hunk 17]
```diff
@@ -157,7 +157,9 @@ public function multiplePayment($clientid,$invoiceChecked, $payment_method,
                     if ($invoice) {
                         if ($total_paid >= $invoice->grand_total) {
                             $invoice_status = 'success';
-                        }
+                        } elseif($total_paid > 0) {
```
Comment (avinash403) [inline app/Http/Controllers/Order/ExtendedBaseInvoiceController.php:160]: which of the above `if` condition will be true when $total_paid is 0? In that case variable `$invoice_status` will be undefined and will throw exception

File: app/Http/Controllers/Order/InvoiceController.php [hunk 18]
```diff
@@ -209,15 +207,21 @@ public function getInvoices(Request $request)
...
+            ->get();
             $user = $this->user->find($invoice->user_id);
             $currency = CartController::currency($user->id);
+            $order = getOrderLink($invoice->order_id);
             $symbol = $currency['symbol'];
 
-            return view('themes.default1.invoice.show', compact('invoiceItems', 'invoice', 'user', 'currency', 'symbol'));
+            return view('themes.default1.invoice.show', compact('invoiceItems', 'invoice', 'user', 'currency', 'symbol','order'));
         } catch (\Exception $ex) {
+            dd($ex);
```
Comment (avinash403) [inline app/Http/Controllers/Order/InvoiceController.php:224]: please remove this

File: app/Http/Controllers/User/ClientController.php [hunk 19]
```diff
@@ -136,6 +124,33 @@ public function getClients(Request $request)
                         ->make(true);
     }
 
+    public function get2faLabel($active)
```
Comment (avinash403) [inline app/Http/Controllers/User/ClientController.php:127]: only class name and title is different. Please create a common method and pass those as params

File: resources/views/vendor/installer/environment-wizard.blade.php (hunk 15 above)
Comment (avinash403) [reply resources/views/vendor/installer/environment-wizard.blade.php:516]: what if someone presses tab?

Comment (avinash403) [reply resources/views/vendor/installer/environment-wizard.blade.php:516]: We should have a backend validation for this instead. This can be bypassed very easily. Someone can copy paste an APP Name with spaces in it

File: app/Http/Controllers/Order/ExtendedBaseInvoiceController.php (hunk 17 above)
Comment (ashu555) [reply app/Http/Controllers/Order/ExtendedBaseInvoiceController.php:160]: It will not be 0. Added validation for the same. 
Reason: When admin goes for creating a payment manually, he cannot create a payment with 0 value

File: app/Http/Controllers/Order/InvoiceController.php (hunk 18 above)
Comment (ashu555) [reply app/Http/Controllers/Order/InvoiceController.php:224]: fixed

File: app/Http/Controllers/User/ClientController.php (hunk 19 above)
Comment (ashu555) [reply app/Http/Controllers/User/ClientController.php:127]: Fixed
